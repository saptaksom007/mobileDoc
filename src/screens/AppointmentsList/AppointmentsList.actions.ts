export const ActionTypes = {
  APPOINTMENTSLIST_REQUESTED: 'APPOINTMENTSLIST_REQUESTED',
  APPOINTMENTSLIST_SUCCEEDED: 'APPOINTMENTSLIST_SUCCEEDED',
  APPOINTMENTSLIST_FAILED: 'APPOINTMENTSLIST_FAILED',
  APPOINTMENTSLIST_LOADED: 'APPOINTMENTSLIST_LOADED',
}

export default class Actions {
  static appointmentsListRequested(addPhysicianAvatars: boolean = true) {
    return {
      type: ActionTypes.APPOINTMENTSLIST_REQUESTED,
      payload: {
        addPhysicianAvatars,
      },
    }
  }
  static appointmentsListSucceeded(rawList: object[]) {
    return {
      type: ActionTypes.APPOINTMENTSLIST_SUCCEEDED,
      payload: {
        rawList,
      },
    }
  }
  static appointmentsListFailed(error: Error) {
    return {
      type: ActionTypes.APPOINTMENTSLIST_FAILED,
      payload: {
        error,
      },
    }
  }
  static appointmentslistLoaded() {
    return {
      type: ActionTypes.APPOINTMENTSLIST_LOADED,
    }
  }
}
