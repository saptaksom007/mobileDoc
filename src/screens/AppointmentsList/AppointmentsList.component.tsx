import React, { Component } from 'react'
import { InteractionManager, Text, TouchableOpacity, View } from 'react-native'
import { connect } from 'react-redux'

import { compose } from 'recompose'
import Conversation from 'components/Conversation/Conversation.component'
import ListContainer from 'components/ListContainer/ListContainer.component'
import withoutKeyboard from 'screens/enhancers/withoutKeyboard'

import { momentWithLocales as moment } from 'utilities/date'

import { findPatientSelf } from 'common-docdok/lib/domain/healthrelation/selectors/findPatient'

import i18n from 'ex-react-native-i18n'
import { Actions as NavAction } from 'navigation/SagaNavigation'
import NotificationsActions from 'notifications/actions'

import { getTestId } from 'utilities/environment'

import Actions from './AppointmentsList.actions'
import styles from './AppointmentsList.styles'
import { Action } from 'redux'
import { filterOldAppointment } from 'api/appointments'
import { Icon } from 'react-native-elements'
import { Color } from 'constants/Color'
import BookAppointmentActions from 'screens/BookAppointment/BookAppointment.actions'

const testId = getTestId('appointmentslist')

interface Props {
  dispatch(action: Action): void
  rawList: any[]
  patientSelf: any
  navigationData: any
}

function getAppointmentContent(appointment: any) {
  switch (appointment.appointmentType) {
    case 'Schedule': {
      return i18n.t('appointmentslist.contentScheduleItem')
    }
    case 'Appointment': {
      const appointmentTime = moment(appointment.appointmentTime)
      return i18n.t('appointmentslist.contentItem', {
        hour: appointmentTime.format('LT'),
        day: appointmentTime.format('dddd'),
        date: appointmentTime.format('Do'),
        month: appointmentTime.format('MMMM'),
      })
    }

    case 'Confirmation': {
      const appointmentTime = moment(appointment.appointmentTime)
      return i18n.t('appointmentslist.contentItem', {
        hour: appointmentTime.format('LT'),
        day: appointmentTime.format('dddd'),
        date: appointmentTime.format('Do'),
        month: appointmentTime.format('MMMM'),
      })
    }

    default: {
      return ''
    }
  }
}

function getAppointmentTitle(appointment: any, patientSelf: any) {
  const name =
    patientSelf && patientSelf.uuid === appointment.healthSubjectUuid
      ? appointment.physician
      : appointment.patientName
  switch (appointment.appointmentType) {
    case 'Schedule': {
      return appointment.clinicName
    }

    case 'Appointment': {
      return `${name} - ${i18n.t('appointmentslist.reminder')}`
    }

    case 'Confirmation': {
      return `${name} - ${i18n.t('appointmentslist.confirmation')}`
    }

    default: {
      return ''
    }
  }
}

export class AppointmentsList extends Component<Props> {
  componentDidMount() {
    InteractionManager.runAfterInteractions(() => {
      const { dispatch, navigationData } = this.props
      console.log('navigation is', JSON.stringify(navigationData))
      dispatch(Actions.appointmentsListRequested())
      dispatch(NotificationsActions.removeNotificationRequested('APPOINTMENT'))
    })
  }

  goToDetail = (appointment: any) => {
    const { dispatch } = this.props
    switch (appointment.appointmentType) {
      case 'Appointment': {
        dispatch(
          NavAction.push('appointment', { appointmentId: appointment.id }),
        )
        return
      }

      case 'Schedule': {
        dispatch(
          NavAction.push('scheduleappointment', {
            appointmentId: appointment.id,
          }),
        )
        return
      }

      case 'Confirmation': {
        dispatch(
          NavAction.push('appointmentconfirmation', {
            appointmentId: appointment.id,
          }),
        )
        return
      }

      default:
        dispatch(NavAction.showLocalError(i18n.t('error.default')))
        break
    }
  }

  renderRow = ({ item: appointment }: any) => {
    const { patientSelf } = this.props
    const title = getAppointmentTitle(appointment, patientSelf)
    const content = getAppointmentContent(appointment)
    return (
      <Conversation
        badgeNumber={appointment.markAsRead ? 0 : 1}
        avatarUrl={appointment.physicianAvatarPicture}
        title={title}
        altAvatar={appointment.physician}
        postedAt={appointment.createdAt}
        iconName='event'
        content={content}
        detailNumberOfLines={3}
        onPress={() => this.goToDetail(appointment)}
      />
    )
  }

  render() {
    const { rawList, dispatch, navigationData } = this.props

    return (
      <>
        <View
          style={[
            styles.padder,
            styles.eventHeader,
            {
              backgroundColor: Color.purple,
            },
          ]}
        >
          <View style={styles.row}>
            <Icon
              name='event'
              color={Color.white}
              containerStyle={{ marginLeft: 12 }}
            />
            <Text style={{ color: Color.white }}>
              {i18n.t('appointmentslist.subtitle')}
            </Text>
          </View>

          <TouchableOpacity
            onPress={() => {
              dispatch(
                BookAppointmentActions.navigateToBookAppointment(
                  navigationData?.forceConversationId ||
                    navigationData?.conversationId,
                  navigationData?.patients?.[0]?.clinicId,
                  navigationData?.scheduleAppointment,
                ),
              )
            }}
            style={[styles.fab]}
          >
            <Icon
              type='ionicon'
              name='add'
              color={Color.purple}
              iconStyle={{ fontWeight: 'bold' }}
            />
          </TouchableOpacity>
        </View>
        <ListContainer
          testID={testId}
          data={rawList}
          renderItem={this.renderRow}
        />
      </>
    )
  }
}

function select({
  appointmentslist: { rawList },
  healthrelation: { patients },
  calendarEvents: { navigationData },
}: any) {
  const patientSelf = patients.find(findPatientSelf)
  return {
    rawList: filterOldAppointment(rawList || []).sort((ap1: any, ap2: any) => {
      if (ap1.createdAt === ap2.createdAt) {
        return 0
      }
      return new Date(ap1.createdAt).getTime() >
        new Date(ap2.createdAt).getTime()
        ? -1
        : 1
    }),
    patientSelf,
    navigationData,
  }
}

export default compose<Props, any>(
  withoutKeyboard,
  connect(select),
)(AppointmentsList)
