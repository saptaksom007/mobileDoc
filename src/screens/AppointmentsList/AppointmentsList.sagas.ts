import { call, put, select, takeEvery } from 'redux-saga/effects'
import { pathOr, filter, last, pipe } from 'ramda'

import { Actions as NavAction } from 'navigation/SagaNavigation'
import i18n from 'ex-react-native-i18n'

import { ActionTypes as LoginActionTypes } from 'screens/Login/Login.actions'
import { asyncAction } from 'common-docdok/lib/utils/sagaUtils'

import { ActionTypes as ScheduleAppointmentActionTypes } from 'screens/ScheduleAppointment/ScheduleAppointment.actions'

import Actions, { ActionTypes } from './AppointmentsList.actions'
import {
  filterOldAppointment,
  fetchAppointmentsListAsync,
} from 'api/appointments'

const isLoggedInSelector = pathOr(false, ['login', 'isLoggedIn'])

const getLatestNotification = pipe(
  pathOr([], ['notifications', 'notificationsList']),
  // @ts-ignore
  filter((notif: any) => notif.origin === 'selected'),
  last,
)

const getAppointmentsList = pipe(
  pathOr([], ['appointmentslist', 'rawList']),
  filterOldAppointment,
)

const getMarkAsReadAppointmentIds = pathOr(
  [],
  ['appointmentslist', 'markAsRead'],
)

const getScheduleAppointments = pipe(
  pathOr([], ['scheduleappointment', 'cacheAppointments']),
)

const noScheduleAppointment = (cacheAppointments: number[]) => (
  appointment: any,
) => !cacheAppointments.includes(appointment.id)

const addMarkAsReadProperty = (app: any[], ids: number[]) =>
  app.reduce(
    (accumulator, current) => [
      ...accumulator,
      ids.includes(current.id)
        ? { ...current, markAsRead: true }
        : { ...current, markAsRead: false },
    ],
    [],
  )

function* fetchAppointmentsList() {
  try {
    const rawList = yield call(fetchAppointmentsListAsync)
    const cacheAppointments: number[] = yield select(getScheduleAppointments)

    if (rawList) {
      const appointmentWithoutSchedule = rawList.filter(
        noScheduleAppointment(cacheAppointments),
      )
      const readIds: number[] = yield select(getMarkAsReadAppointmentIds)

      const appointmentList = addMarkAsReadProperty(
        appointmentWithoutSchedule,
        readIds.concat(cacheAppointments),
      )
      yield put(Actions.appointmentsListSucceeded(appointmentList))
    } else {
      yield put(Actions.appointmentsListSucceeded([]))
    }
  } catch (err) {
    yield put(Actions.appointmentsListFailed(err))
    yield put(
      NavAction.showLocalError(
        i18n.t('appointmentslist.errors.fetchAppointmentsListAsync'),
        'login',
      ),
    )
  }
}

function* fetchAppointmentsListOnStartup() {
  try {
    yield asyncAction(Actions.appointmentsListRequested())
    const rawList = yield select(getAppointmentsList)
    if (rawList && rawList.length > 0) {
      const isLoggedIn = yield select(isLoggedInSelector)
      if (!isLoggedIn) {
        const latestNotification = yield select(getLatestNotification)
        if (
          !latestNotification ||
          (latestNotification &&
            latestNotification.addedAt > Date.now() + 10000)
        ) {
          yield put(NavAction.push('Appointment'))
        }
      }
    }
  } catch (err) {
    yield put(Actions.appointmentsListFailed(err))
    yield put(
      NavAction.showLocalError(
        i18n.t('appointmentslist.errors.fetchAppointmentsListAsync'),
        'login',
      ),
    )
  }
}

export default [
  takeEvery(
    LoginActionTypes.GET_LIMITED_ACCESS_TOKEN_SUCCEEDED,
    fetchAppointmentsListOnStartup,
  ),
  takeEvery(ActionTypes.APPOINTMENTSLIST_REQUESTED, fetchAppointmentsList),
  takeEvery(
    ScheduleAppointmentActionTypes.SCHEDULE_APPOINTMENT,
    fetchAppointmentsList,
  ),
]
