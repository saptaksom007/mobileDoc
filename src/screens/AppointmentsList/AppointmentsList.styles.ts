import { Color } from 'constants/Color'
import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    justifyContent: 'center',
    alignItems: 'center',
  },
  eventHeader: {
    marginTop: -8,
    height: 60,
  },
  padder: {
    padding: 12,
    marginBottom: 20,
  },
  fab: {
    elevation: 5,
    position: 'absolute',
    top: 18,
    left: 24,
    height: 60,
    width: 60,
    backgroundColor: Color.white,
    borderRadius: 30,
    justifyContent: 'center',
    alignItems: 'center',
    shadowColor: 'black',
    shadowOpacity: 0.1,
    shadowRadius: StyleSheet.hairlineWidth,
    shadowOffset: {
      width: StyleSheet.hairlineWidth,
      height: StyleSheet.hairlineWidth,
    },
  },
  row: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    paddingTop: 12,
  },
})
