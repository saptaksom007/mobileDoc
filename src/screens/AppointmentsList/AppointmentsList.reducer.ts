import { uniq } from 'ramda'
import { ActionTypes } from './AppointmentsList.actions'
import { ActionTypes as AppointmentsConfirmationActionTypes } from '../AppointmentConfirmation/AppointmentConfirmation.actions'

const {
  APPOINTMENTSLIST_REQUESTED,
  APPOINTMENTSLIST_SUCCEEDED,
  APPOINTMENTSLIST_FAILED,
  APPOINTMENTSLIST_LOADED,
} = ActionTypes

interface State {
  error?: Error
  rawList: object[]
  loaded: boolean
  markAsRead: number[]
}

interface Action {
  type: keyof typeof ActionTypes
  payload: {
    error?: Error
    rawList?: object[]
    reloadNavigation?: boolean
  }
}

const initialState: State = {
  error: undefined,
  rawList: [],
  loaded: false,
  markAsRead: [],
}

export default function appointmentslist(
  state: State = initialState,
  action: Action,
) {
  switch (action.type) {
    case APPOINTMENTSLIST_REQUESTED: {
      return {
        ...state,
        error: undefined,
      }
    }

    case APPOINTMENTSLIST_SUCCEEDED: {
      const { rawList } = action.payload
      return {
        ...state,
        error: undefined,
        rawList,
        loaded: true,
      }
    }

    case APPOINTMENTSLIST_FAILED: {
      return {
        ...state,
        error: action.payload.error,
        rawList: [],
        loaded: true,
      }
    }

    case APPOINTMENTSLIST_LOADED: {
      return {
        ...state,
        loaded: true,
      }
    }

    // tslint:disable-next-line:max-line-length
    case AppointmentsConfirmationActionTypes.MARK_AS_READ_APPOINTMENT_CONFIRMATION: {
      const readAppointment = {
        ...state.rawList.find((app: any) => app.id === action.payload),
        markAsRead: true,
      }
      return {
        ...state,
        rawList: [
          ...state.rawList.filter((app: any) => app.id !== action.payload),
          readAppointment,
        ],
        markAsRead: uniq([...state.markAsRead, action.payload]),
      }
    }

    default:
      return state
  }
}
