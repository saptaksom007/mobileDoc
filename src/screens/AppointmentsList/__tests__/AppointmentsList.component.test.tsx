import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { AppointmentsList } from '../AppointmentsList.component'

it('<AppointmentsList/> render correctly', () => {
  const tree = renderer
    .create(
      <AppointmentsList
        rawList={[]}
        patientSelf={{}}
        dispatch={() => {}}
        navigationData={{}}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
