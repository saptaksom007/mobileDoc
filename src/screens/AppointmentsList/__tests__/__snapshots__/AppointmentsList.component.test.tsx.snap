// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppointmentsList/> render correctly 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "marginBottom": 20,
          "padding": 12,
        },
        Object {
          "height": 60,
          "marginTop": -8,
        },
        Object {
          "backgroundColor": "#8950b6",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingTop": 12,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
          }
        }
      >
        appointmentslist.subtitle
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 30,
          "elevation": 5,
          "height": 60,
          "justifyContent": "center",
          "left": 24,
          "opacity": 1,
          "position": "absolute",
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 0.5,
            "width": 0.5,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 0.5,
          "top": 18,
          "width": 60,
        }
      }
    />
  </View>,
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 1294,
        "justifyContent": "center",
      }
    }
  >
    <Text
      allowFontScaling={false}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Array [
          Object {
            "fontFamily": "TitilliumWeb-Regular",
            "fontSize": 17.5,
            "lineHeight": 22,
            "overflow": "visible",
          },
          Object {
            "color": "rgba(27, 106, 158, 0.85)",
            "fontSize": 17.5,
            "lineHeight": 21,
          },
          Object {
            "color": "#999",
            "marginTop": -40,
          },
        ]
      }
    >
      global.empty
    </Text>
  </View>,
]
`;
